### PR TITLE
[Feature][Model] Enable DeepSeek-V4.1-Flash + DSpark on Ascend950 (A5)

### PR DESCRIPTION
**Pull request by yiminghub2024**
### What this PR does / why we need it?

Enables **DeepSeek-V4.1-Flash + DSpark on Ascend950 (A5)**.

Before this change the model starts but is unusable: generation is garbled and loops, and DSpark accepts only the first speculative position.

#### Root cause of the garbled output

The A5 `HcPre` kernel ignores the `pre_mix` weights it is given and collapses the four hyper-connection streams by its own internally computed coefficients. At layer 0 all four streams are the token embedding, so the layer input becomes

```
y = (sum of the kernel's own coefficients) x embedding = 0.356 x embedding
```

instead of `1.0 x embedding`. Every layer is fed its input scaled down by ~2.8x, and the error compounds through the stack. The same kernel also leaves its optional mixing-coefficient output unwritten.

`DeepseekV41DecoderLayer.hc_pre` now performs both reductions itself. The combination is the same weighted sum `hc_collapse` already implements, so no numerical behaviour is invented:

```python
y = (pre_mix.unsqueeze(-1) * x.float()).sum(-2).to(x.dtype)
```

DSpark is fixed by the same change: the draft model consumes the main model's hidden states, so once those were wrong the draft could not match beyond the first position.

#### The rest of the A5 enablement

| Area | Change |
| --- | --- |
| Attention | Use `SparseFlashMla` on A5 instead of `npu_sparse_attn_sharedkv`, with `PA_BBND` layout and sliding-window `ori` masking |
| CSA odd-length trap | Duplicate one query per request so no request ever presents an odd packed query count (AI core 507015) |
| Compressed KV geometry | Derive the storage block size from `tokens_per_state`; a compress-ratio-2 plane stores half a logical block per page |
| MoE dispatch | Fall back to AllGather on A5, where `aclnnMoeDistributeDispatchV4` fails with 561000 (gated on `DSV4_COMPRESSED_CACHE`, other SoCs unaffected) |
| Quantization | Select the FP8 scheme from the checkpoint's `weight_block_size`; released V4.1-Flash ships 32-wide blocks and must skip the 128x128 to 32x1 scale reorganization |
| Indexer | Select TopK straight from the compressed stream: `quant_lightning_indexer_v2` only implements the two-level candidate path in its arch22 kernel |
| Plan construction | Resolve `npu_sparse_attn_sharedkv` at call time, since V4.1 builds the DSA plan while configuring the KV cache |
| Custom ops | Add `sparse_flash_mla`, `sparse_flash_mla_metadata` and `rms_norm_cast` to the ascend950 build list |

### Does this PR introduce any user-facing change?

No new flags or configuration. DeepSeek-V4.1-Flash simply works on A5.

### How was this patch tested?

Hardware: Ascend950PR, 8 cards, TP8 + EP, released `DeepSeek-V4.1-Flash` weights.

```bash
vllm serve /data/DeepSeek-V4.1-Flash \
  --served-model-name dsv41 --host 0.0.0.0 --port 8000 \
  --tensor-parallel-size 8 --enable-expert-parallel \
  --block-size 128 --max-model-len 1000000 \
  --max-num-batched-tokens 4096 --max-num-seqs 10 \
  --gpu-memory-utilization 0.92 \
  --no-disable-hybrid-kv-cache-manager \
  --compilation-config '{"cudagraph_mode":"FULL_DECODE_ONLY","max_cudagraph_capture_size":64}' \
  --tokenizer-mode deepseek_v41 --reasoning-parser deepseek_v41 \
  --tool-call-parser deepseek_v41 --trust-remote-code \
  --safetensors-load-strategy lazy \
  --quantization deepseek_v4_fp8 \
  --speculative-config '{"method":"dspark","num_speculative_tokens":5,"enforce_eager":true}'
```

**Generation quality.** A 1M-token context serves with the KV pool clamped to 9,592,505 tokens (9.59x concurrency for 1M-token requests). An 800-word Chinese essay prompt finishes on its own at 1247 tokens, 50.9 tok/s, with every date and name correct. Mixed long/short/long requests answer correctly and hit the prefix cache (0.2s on the repeated 532-token prompt).

**Dataset accuracy (GSM8K).** 5-shot GSM8K over `/v1/chat/completions`, `temperature=0`, thinking disabled, answer taken as the last number in the reply, scored against the `####` gold value:

| Setting | Samples | exact_match |
| --- | --- | --- |
| 5-shot, one request in flight | 200 | **190 / 200 = 95.0%** |
| 5-shot, one request in flight | 50 | 50 / 50 = 100% |

One of the ten misses is a formatting difference only (`57.00` against a gold of `57`). Two replies out of 200 hit the 512-token cap; the mean reply is 121 tokens, so generation terminates on its own.

This was measured on the command above with `--max-model-len 69632` and `--no-enable-prefix-caching`, and with requests issued one at a time.

**Known limitation: batched decode.** Accuracy is currently tied to the number of requests in flight. On the same server process that scored 50/50 sequentially, driving the same 50 questions at 8 concurrent requests scores 2/50, with 42 of the 50 replies running to the token cap and one degenerating into a run of repeated `0` digits. Turning prefix caching off does not change this (4% off, 6% on), and it reproduces within a single process, so it is neither prefix caching nor a start-up race: something in the multi-request decode path is corrupting per-request state. Single-stream serving is correct and is what the numbers above cover; multi-request throughput serving is not fixed by this PR, and I am treating it as the next piece of work rather than claiming it here.

**Agreement with a known-good reference.** Compared against a working DeepSeek-V4.1-Flash deployment on the same weights:

| Metric | Before | After |
| --- | --- | --- |
| Layers with identical routed-expert sets | 3/40 | 30/40 |
| Mean expert overlap | 3.0/6 | 5.72/6 |
| DSpark per-position acceptance | 0.135, 0, 0, 0, 0 | 0.61, 0.35, 0.17, 0.09, 0.05 |
| Mean acceptance length | ~1.1 | 1.8 - 2.3 |

Short-prompt logit distributions now match the reference to within bf16 noise, including top-5 ordering.

**Unit tests.** `tests/ut` was run against this branch and against unmodified main in the same container to separate pre-existing failures (missing `acl` module and LoRA ops in that image) from real regressions. Test updates in this PR cover the A5 behaviour changes: the `hc_pre` contract, A5 MoE dispatch collapsing to AllGather, and metadata fixtures gaining `num_reqs`.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/attention/dsa_v41.py [hunk 1]
```diff
@@ -509,15 +730,20 @@ def _forward_attention(self, attn, q, metadata, compressed_indices, *, source_ca
...
             layout_q="TND",
             layout_kv="PA_BBND",
             topk_value_mode=1,
             return_softmax_lse=False,
         )
+        if even_plan is not None:
+            output = output.index_select(0, even_plan["keep"])
+        if keep is not None:
+            idx = torch.tensor(keep, device=output.device, dtype=torch.long)
+            output = output.index_select(0, idx)
```
[the remark this answers is not in the document]
Comment (yiminghub2024) [reply vllm_ascend/attention/dsa_v41.py]: Good catch, applied. I took it one step further than the suggestion: since `keep` only ever carried `list(range(orig_rows))`, it now carries `orig_rows` itself, so there is no list to allocate or measure at all:

```python
        if keep is not None:
            # The padding is appended, so the real rows are a prefix.
            output = output[:keep]
```

The odd-prefill pad is always *appended* (`torch.cat([hidden_states, hidden_states[-1:].expand(extra, ...)])`), so the rows to keep are exactly a prefix and the slice is equivalent to the old `index_select`. Pushed in e6f06e0.

File: vllm_ascend/attention/dsa_v41.py [hunk 2]
```diff
@@ -216,10 +297,93 @@ def scatter_cache_sk(
...
+    except RuntimeError:
+        # V4.1 SWA/long-KV are views into a larger layer slot, so the page
+        # stride is not head_dim+128. Quantize into a packed scratch, then
+        # store through the stride-aware scatter.
+        packed = torch.zeros(
+            (values.shape[0], 1, cache.shape[-1]),
+            dtype=cache.dtype,
+            device=cache.device,
+        )
+        seq = torch.arange(values.shape[0], device=values.device, dtype=torch.int32)
```
[the remark this answers is not in the document]
Comment (yiminghub2024) [reply vllm_ascend/attention/dsa_v41.py:343]: You are right that this is the branch V4.1 SWA/long-KV actually take, so the per-step `torch.zeros` + `torch.arange` is real work rather than a rare fallback. Two corrections on the impact, then what I plan to do.

On graph capture: this PR runs A5 decode with `cudagraph_mode: FULL_DECODE_ONLY` and capture succeeds today. Inside a captured decode the row count is static, so these allocations are taken from the graph's memory pool and baked in rather than blocking capture. The cost that does remain is the eager/prefill path, where it is a per-layer, per-step allocation.

On fixing it here: `store_attn_kv` is on the KV store path, which is where most of the correctness bugs in this enablement lived (wrong `storage_block_size` silently shifting compressed-KV page offsets, and the SWA/long-KV views this branch exists for). It is also a free function that currently receives only the cache, the values and the slot mapping, so pre-allocating means plumbing a max-rows bound and an owner for the scratch buffer through to it. That is a mechanical change, but on this path it needs a full A5 re-validation rather than review-time confidence, so I would rather not fold it into this PR.

I will send it as a follow-up once this lands, so the buffer change is reviewable on its own with its own hardware run. If you would prefer it inside this PR instead, say so and I will do it here.

File: vllm_ascend/ops/fused_moe/shared_experts.py [hunk 3]
```diff
@@ -545,7 +548,12 @@ def _select_mlp_path(self) -> SharedExpertMLPPath:
         if has_quantized_shared_without_lora and self.quant_type in (QuantType.W8A8, QuantType.W4A8):
             return SharedExpertMLPPath.A8_INT_FUSED
         if has_quantized_shared_without_lora and self.quant_type == QuantType.W4A8MXFP:
-            return SharedExpertMLPPath.A8_MXFP_FUSED
+            # The fused activation quant reads 128-wide groups, so a shared
+            # intermediate that TP splits below that alignment must stay on the
+            # wrapper path.
+            gate_up_width = getattr(self.layer.gate_up_proj, "output_size_per_partition", None)
```
[the remark this answers is not in the document]
Comment (yiminghub2024) [reply vllm_ascend/ops/fused_moe/shared_experts.py:554]: I do not think the guard would buy anything here, because control cannot reach this line with a `layer` that lacks `gate_up_proj`.

The predicate one branch above (pre-existing, not part of this PR) already dereferences it unguarded:

```python
            and hasattr(self.layer.gate_up_proj, "weight_scale")   # line 545
```

and it is a precondition of `has_quantized_shared_without_lora`, which this branch requires. The execution paths assume it too: `self.layer.gate_up_proj(hidden_states)` in `part1`, plus `.weight` / `.weight_scale` / `.weight_scale_fp32` in the fused paths. So a shared-expert MLP without `gate_up_proj` would already have raised at line 545, and adding `getattr` only at 554 would convert a real misconfiguration into a silent path switch rather than handle it.

The case that can genuinely occur is the attribute existing without an integer width, e.g. a linear that was never TP-partitioned, and that is what the `not isinstance(gate_up_width, int)` check covers: it falls through to `A8_MXFP_FUSED` rather than trusting a non-int. Happy to revisit if you have a concrete architecture in mind that reaches `W4A8MXFP` shared experts without a `gate_up_proj`.

**Comment by yiminghub2024** [comment]
CI note for maintainers: `Recommend tests` is failing for a reason outside this PR, and it takes `select-tests` and `ci-gate` down with it.

```
Fetching changes from GitHub PR: vllm-project/vllm-ascend#17122
  Attempt 1/3 to get PR diff via GitHub API...
  Attempt 1 failed: HTTP Error 503: Service Unavailable
  ... (3/3 failed)
```

`test_selector.py` requests `https://api.github.com/repos/{repo}/pulls/{num}` from the self-hosted runner and gets a 503 on all three attempts. The same request succeeds from outside the runner (the diff is served fine, 32 files / +1427 / -238, and the PR reports `mergeable: true`), so this looks like runner egress to `api.github.com` rather than anything about the diff. It reproduced on two separate runs, so it is not a one-off.

Everything else is green: `DCO`, `pre-commit`, `cpu-ut`, `main`, `PR create action`.

I cannot re-run jobs or change labels on this repo. Either a re-run once runner networking is healthy, or swapping `ready-precise` for the all-tests label (which skips the recommendation step), would unblock the gate. Happy to do whatever is easiest on your side.

**Comment by yiminghub2024** [comment]
/rerun
[Bot]: rerun completed.

**Rerun (failed jobs only):**
- E2E

**Comment by yiminghub2024** [comment]
https://github.com/vllm-project/recipes/pull/1028

**Comment by yiminghub2024** [comment]
@wangxiyuan hi ,please take a look ablout this ,thanks ~
